back/app/services/projects_finder_admin_service.rb

```diff
@@ -0,0 +1,144 @@
+
+    # I first did this with a group by and max, but Copilot suggested
+    # using a window function instead, which is more efficient
+    recent_pageviews_sql = <<-SQL.squish
+      SELECT
+        #{substring_statement}::UUID AS admin_project_id,
+        impact_tracking_pageviews.created_at AS last_viewed_at,
+        ROW_NUMBER() OVER (
+          PARTITION BY #{substring_statement}
+          ORDER BY impact_tracking_pageviews.created_at DESC
+        ) AS rn
+      FROM impact_tracking_pageviews
+      INNER JOIN impact_tracking_sessions
+        ON impact_tracking_pageviews.session_id = impact_tracking_sessions.id
+      WHERE path LIKE '%admin/projects/%'
+        AND impact_tracking_sessions.user_id = '#{current_user.id}'
+    SQL
+
+    recent_pageviews_subquery = "(#{recent_pageviews_sql}) AS recent_pageviews"
+
+    projects_subquery = scope
+      .joins("LEFT JOIN #{recent_pageviews_subquery} ON recent_pageviews.admin_project_id = projects.id AND recent_pageviews.rn = 1")
+      .select('recent_pageviews.last_viewed_at AS last_viewed_at, projects.*')
+
+    # We order by last_viewed_at, but tie-break with created_at and id for a stable sort,
+    # which is important for pagination
+    Project
+      .from(projects_subquery, :projects)
+      .order('last_viewed_at DESC NULLS LAST, projects.created_at ASC, projects.id ASC')
+  end
+
+  def self.sort_phase_starting_or_ending_soon(scope)
+    phases_ending_soon_subquery = Phase
+      .where("coalesce(end_at, 'infinity'::DATE) >= current_date")
+      .group(:project_id)
+      .select("project_id, min(coalesce(end_at, 'infinity'::DATE)) AS min_end_at")
+
+    phases_starting_soon_subquery = Phase
+      .where('start_at >= current_date')
+      .group(:project_id)
+      .select('project_id, min(start_at) AS min_start_at')
+
+    projects_subquery = scope
+      .joins("LEFT JOIN (#{phases_ending_soon_subquery.to_sql}) AS phases_ending_soon ON phases_ending_soon.project_id = projects.id")
+      .joins("LEFT JOIN (#{phases_starting_soon_subquery.to_sql}) AS phases_starting_soon ON phases_starting_soon.project_id = projects.id")
+      .select('least(phases_ending_soon.min_end_at, phases_starting_soon.min_start_at) AS soon_date, projects.*')
+
+    # We order by soon_date, but tie-break with created_at and id for a stable sort,
+    # which is important for pagination
+    Project
+      .from(projects_subquery, :projects)
+      .order('soon_date ASC NULLS LAST, projects.created_at ASC, projects.id ASC')
+  end
+
+  # FILTERING METHODS
+  def self.filter_status(scope, params = {})
+    status = params[:status] || []
+
+    if status.present?
+      scope
+        .joins("INNER JOIN admin_publications ON admin_publications.publication_id = projects.id AND admin_publications.publication_type = 'Project'")
+        .where(admin_publications: { publication_status: status })
+    else
+      scope
+    end
+  end
```

Comment:
I prefer this more compact way, when we just return the scope unchanged. Think it makes the code more readable. Comment applies to the other methods too.

```suggestion
  def self.filter_status(scope, params = {})
    status = params[:status] || []
    return scope unless status.present?

    scope
      .joins("INNER JOIN admin_publications ON admin_publications.publication_id = projects.id AND admin_publications.publication_type = 'Project'")
      .where(admin_publications: { publication_status: status })
  end
```